server/service/hosts.go

```diff
@@ -893,3 +893,33 @@ func hostsReportEndpoint(ctx context.Context, request interface{}, svc fleet.Ser
 	}
 	return hostsReportResponse{Hosts: hosts}, nil
 }
+
+type osVersionsRequest struct {
+	TeamID   *uint   `query:"team_id,optional"`
+	Platform *string `query:"platform,optional"`
+}
+
+func osVersionsEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (interface{}, error) {
+	req := request.(*osVersionsRequest)
+
+	osVersions, err := svc.OSVersions(ctx, req.TeamID, req.Platform)
```

Comment:
You have to declare a corresponding response struct for the endpoint, e.g. like for this other endpoint:

```
type getAggregatedMacadminsDataResponse struct {
	Err       error                          `json:"error,omitempty"`
	Macadmins *fleet.AggregatedMacadminsData `json:"macadmins"`
}
func (r getAggregatedMacadminsDataResponse) error() error { return r.Err }
```

And then you set the result of that response on success, and the error on failure, and always return that response and nil as error, e.g. again from that other endpoint:

```
        if err != nil {
		return getAggregatedMacadminsDataResponse{Err: err}, nil
	}
	return getAggregatedMacadminsDataResponse{Macadmins: data}, nil
```

Reply:
Regarding the need to implement the `error() error` method, this is an annoyance I hope we can remove, added it to this ticket: https://github.com/fleetdm/fleet/issues/4406#issuecomment-1076384407